compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs

```diff
@@ -247,6 +247,27 @@ infoStep world e = Info
     (safetyStep (fmap safety e))
     (typeclassStep world (fmap tcinfo e))
 
+-- | Take the free variables and safety of a let-expression `let x = e1 in e2`
+-- and compute over-approximations of the free variables and
+-- under-approximations of the safe of `e1` and `e2`. The reasoning behind the
+-- choice of `s1` and `s2` is as follows:
+-- * If `fv(let x = e1 in e2) ⊆ V`, then `fv(e1) ⊆ V` and `fv(e2) ⊆ V ∪ {x}`.
+-- * If `let x = e1 in e2` is k-safe, then `e1` is 0-safe and `e2` is k-safe.
+infoUnstepELet :: ExprVarName -> Info -> (Info, Info)
+infoUnstepELet x (Info fv sf _) = (s1, s2)
+  where
+    s1 = Info fv (sf `min` Safe 0) TCNeither
```

Comment:
There are no silly questions on such complicated matters!

The expression `1` is 0-safe because it does not raise any exceptions. (We do not consider `k`-safe for any `k > 0` because you cannot apply arguments to it.) The expression `1 + error "boom"` is unsafe and not 1-safe because it raises an exception. The difference between these two expressions is that it's ok for the simplifier to delete the former when it's dead code but not the latter (because of strictness).

Reply:
Oh that makes sense, I think I got confused by looking at the code and seeing `BEError` marked as 0-safe and was thinking of `error "abc"` instead of `error`. Thanks for the explanation!